# HW5, Myunghee Lee (USC ID: 2446752777)

1. Multi-class and Multi-Label Classification Using Support Vector Machines

(a) Choose 70% of the data randomly as the training set. 

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split  # to divide training and test data

df = pd.read_csv("Frogs_MFCCs.csv")
print(df)

      MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  MFCCs_ 4  MFCCs_ 5  MFCCs_ 6  MFCCs_ 7  \
0          1.0  0.152936 -0.105586  0.200722  0.317201  0.260764  0.100945   
1          1.0  0.171534 -0.098975  0.268425  0.338672  0.268353  0.060835   
2          1.0  0.152317 -0.082973  0.287128  0.276014  0.189867  0.008714   
3          1.0  0.224392  0.118985  0.329432  0.372088  0.361005  0.015501   
4          1.0  0.087817 -0.068345  0.306967  0.330923  0.249144  0.006884   
5          1.0  0.099704 -0.033408  0.349895  0.344535  0.247569  0.022407   
6          1.0  0.021676 -0.062075  0.318229  0.380439  0.179043 -0.041667   
7          1.0  0.145130 -0.033660  0.284166  0.279537  0.175211  0.005791   
8          1.0  0.271326  0.027777  0.375738  0.385432  0.272457  0.098192   
9          1.0  0.120565 -0.107235  0.316555  0.364437  0.307757  0.025992   
10         1.0  0.148539 -0.096910  0.257523  0.260881  0.312603  0.134134   
11         1.0  0.277948  0.091657  0.331656  0.307372  0.257359

In [2]:
X = df.loc[:, :'MFCCs_22'] # independent variables
labels = df.loc[:, 'Family':]   # lables: Family, Genus, and Species

# Choose 70% of the data randomly as the training set
X_train, X_test, labels_train, labels_test = train_test_split(X, labels, test_size=0.3, random_state=15)

# 1. Family
y1_train = labels_train.loc[:,'Family']
y1_test = labels_test.loc[:,'Family']

# 2. Genus
y2_train = labels_train.loc[:,'Genus']
y2_test = labels_test.loc[:,'Genus']

# 3. Species
y3_train = labels_train.loc[:,'Species']
y3_test = labels_test.loc[:,'Species']

1. (b) Each instance has three labels: Families, Genus, and Species. Each of the labels has multiple classes. We wish to solve a multi-class and multi-label problem. One of the most important approaches to multi-class classification is to train a classifier for each label. We first try this approach: 

i. Research exact match and hamming score/ loss methods for evaluating multilabel classification and use them in evaluating the classifiers in this problem.

(Answer) Exact match is the percentage of instances having all their labels classified correctly. 

Hamming score is the fraction of correctly classifed labels to the total number of labels. 

Hamming loss is the fraction of wrongly classified labels to the total number of labels. Thus, hamming score is '1-Hamming loss.'

1. (b) ii. Train a SVM for each of the labels, using Gaussian kernels and one versus all classifiers. Determine the weight of the SVM penalty and the width of the Gaussian Kernel using 10 fold cross validation.1 You are welcome to try to solve the problem with both standardized 2 and raw attributes and report the results. 

In [3]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
import numpy as np

# decision_function_shape='ovr': one vesus rest
# kernel='rbf': Gaussian kernel
clf = SVC(kernel='rbf', decision_function_shape='ovr')

# C: the weight of the SVM penalty
# gamma: the width of the Gaussian Kernel
C_range = np.logspace(-2, 2, 5) # 0.01, 0.1, 1, 10, 100
gamma_range = np.logspace(-2, 2, 5) 
param_grid = dict(gamma=gamma_range, C=C_range)
# 10 fold cross validation
grid = GridSearchCV(clf, param_grid=param_grid, cv=10)

In [4]:
# 1. Family (raw data)
# Determine C and gamma by 10 fold CV
grid.fit(X_train, y1_train)
print(grid.best_params_, grid.best_score_)

{'C': 10.0, 'gamma': 1.0} 0.9912629070691025


In [5]:
C=grid.best_params_['C']
gamma=grid.best_params_['gamma']
clf = SVC(C=C, gamma=gamma, kernel='rbf', decision_function_shape='ovr')
clf.fit(X_train, y1_train)
print("train accuaracy: %0.4f" % clf.score(X_train, y1_train))
print("test accuaracy: %0.4f" % clf.score(X_test, y1_test))

train accuaracy: 0.9986
test accuaracy: 0.9893


In [6]:
from sklearn.metrics import hamming_loss
 
y1_pred = clf.predict(X_test)
h1 = hamming_loss(y1_test, y1_pred)
print(h1)  # the fraction of wrong labels for Family

0.010653080129689671


In [7]:
# 2. Genus (raw data)
# Determine C and gamma by 10 fold CV
grid.fit(X_train, y2_train)
print(grid.best_params_, grid.best_score_)

{'C': 10.0, 'gamma': 1.0} 0.9892772041302621


In [8]:
C=grid.best_params_['C']
gamma=grid.best_params_['gamma']
clf = SVC(C=C, gamma=gamma, kernel='rbf', decision_function_shape='ovr')
clf.fit(X_train, y2_train)
print("train accuaracy: %0.4f" % clf.score(X_train, y2_train))
print("test accuaracy: %0.4f" % clf.score(X_test, y2_test))

train accuaracy: 0.9990
test accuaracy: 0.9889


In [9]:
y2_pred = clf.predict(X_test)
h2 = hamming_loss(y2_test, y2_pred)
print(h2) # the fraction of wrong labels for Genus

0.0111162575266327


In [10]:
# 3. Species (raw data)
# Determine C and gamma by 10 fold CV
grid.fit(X_train, y3_train)
print(grid.best_params_, grid.best_score_)

{'C': 10.0, 'gamma': 1.0} 0.9882843526608419


In [11]:
C=grid.best_params_['C']
gamma=grid.best_params_['gamma']
clf = SVC(C=C, gamma=gamma,kernel='rbf', decision_function_shape='ovr')
clf.fit(X_train, y3_train)
print("train accuaracy: %0.4f" % clf.score(X_train, y3_train))
print("test accuaracy: %0.4f" % clf.score(X_test, y3_test))

train accuaracy: 0.9988
test accuaracy: 0.9893


In [12]:
y3_pred = clf.predict(X_test)
h3 = hamming_loss(y3_test, y3_pred)
print(h3) # the fraction of wrong labels for Species

0.010653080129689671


In [13]:
# exact match
# the percentage of instances having all their labels classified correctly

count = 0
for i in range(len(y1_test)):
    if y1_pred[i]==y1_test.iloc[i] and \
    y2_pred[i]==y2_test.iloc[i] and \
    y3_pred[i]==y3_test.iloc[i]:
        count += 1

e =(count*100)/len(y1_test) # exact math
h=(h1+h2+h3)/3 # hamming loss

print("exact match: %0.3f" % e, "%")        
print("hamming loss: %0.3f" % h)

exact match: 98.518 %
hamming loss: 0.011


In [14]:
# standardization
from sklearn.preprocessing import StandardScaler

scX_train = StandardScaler().fit_transform(X_train)
scX_test = StandardScaler().fit_transform(X_test)

In [15]:
# 1. Family (standardized data)
# Determine C and gamma by 10 fold CV
grid.fit(scX_train, y1_train)
print(grid.best_params_, grid.best_score_)

{'C': 10.0, 'gamma': 0.1} 0.9900714853057982


In [17]:
C=grid.best_params_['C']
gamma=grid.best_params_['gamma']
clf = SVC(C=C, gamma=gamma, kernel='rbf', decision_function_shape='ovr')
clf.fit(scX_train, y1_train)
print("train accuaracy: %0.4f" % clf.score(scX_train, y1_train))
print("test accuaracy: %0.4f" % clf.score(scX_test, y1_test))

train accuaracy: 1.0000
test accuaracy: 0.9903


In [18]:
y1_pred = clf.predict(scX_test)
h1 = hamming_loss(y1_test, y1_pred)
print(h1)  # the fraction of wrong labels for Family

0.009726725335803613


In [19]:
# 2. Genus (standardized data)
# Determine C and gamma by 10 fold CV
grid.fit(scX_train, y2_train)
print(grid.best_params_, grid.best_score_)

{'C': 10.0, 'gamma': 0.01} 0.988482922954726


In [20]:
C=grid.best_params_['C']
gamma=grid.best_params_['gamma']
clf = SVC(C=C, gamma=gamma, kernel='rbf', decision_function_shape='ovr')
clf.fit(scX_train, y2_train)
print("train accuaracy: %0.4f" % clf.score(scX_train, y2_train))
print("test accuaracy: %0.4f" % clf.score(scX_test, y2_test))

train accuaracy: 0.9958
test accuaracy: 0.9852


In [21]:
y2_pred = clf.predict(scX_test)
h2 = hamming_loss(y2_test, y2_pred)
print(h2)  # the fraction of wrong labels for Genus

0.014821676702176934


In [22]:
# 3. Species (standardized data)
# Determine C and gamma by 10 fold CV
grid.fit(scX_train, y3_train)
print(grid.best_params_, grid.best_score_)

{'C': 10.0, 'gamma': 0.01} 0.9882843526608419


In [24]:
C=grid.best_params_['C']
gamma=grid.best_params_['gamma']
clf = SVC(C=C, gamma=gamma, kernel='rbf', decision_function_shape='ovr')
clf.fit(scX_train, y3_train)
print("train accuaracy: %0.4f" % clf.score(scX_train, y3_train))
print("test accuaracy: %0.4f" % clf.score(scX_test, y3_test))

train accuaracy: 0.9968
test accuaracy: 0.9889


In [25]:
y3_pred = clf.predict(scX_test)
h3 = hamming_loss(y3_test, y3_pred)
print(h3)  # the fraction of wrong labels for Species

0.0111162575266327


In [26]:
# exact match

count = 0
for i in range(len(y1_test)):
    if y1_pred[i]==y1_test.iloc[i] and \
    y2_pred[i]==y2_test.iloc[i] and \
    y3_pred[i]==y3_test.iloc[i]:
        count += 1

e =(count*100)/len(y1_test) # exact math
h=(h1+h2+h3)/3 # hamming loss

print("exact match: %0.3f" % e, "%")        
print("hamming loss: %0.3f" % h)

exact match: 97.869 %
hamming loss: 0.012


(Answer of 1. (b) ii)
1. raw data

exact match: 98.518 %

hamming loss: 0.011

2. standardized data

exact match: 97.869 %

hamming loss: 0.012

The performance of the raw data is sligthly better than that of the standardized data.

1. (b) iii. Repeat 1(b)ii with L1-penalized SVMs.3 Remember to standardize4 the attributes. Determine the weight of the SVM penalty using 10 fold cross validation. 

In [27]:
from sklearn.svm import LinearSVC

clf = LinearSVC(penalty='l1', multi_class='ovr', dual=False)
C_range = np.logspace(-2, 5, 8) # 10^-2, 10^-1, 1, 10, ...10^5
param_grid = dict(C=C_range)

grid = GridSearchCV(clf, param_grid=param_grid, cv=10)

In [28]:
# 1. Family (L1-penalized SVM)
# Determine C by 10 fold CV
grid.fit(scX_train, y1_train)
print(grid.best_params_, grid.best_score_)

C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of

C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of

{'C': 100.0} 0.9418189038919778


C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [29]:
clf = LinearSVC(C=grid.best_params_['C'], penalty='l1', multi_class='ovr', dual=False)
clf.fit(scX_train, y1_train)
print("train accuaracy: %0.4f" % clf.score(scX_train, y1_train))
print("test accuaracy: %0.4f" % clf.score(scX_test, y1_test))

train accuaracy: 0.9432
test accuaracy: 0.9264


C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [30]:
y1_pred = clf.predict(scX_test)
h1 = hamming_loss(y1_test, y1_pred)
print(h1)

0.07364520611394164


In [31]:
# 2. Genus (L1-penalized SVM)
# Determine C by 10 fold CV
grid.fit(scX_train, y2_train)
print(grid.best_params_, grid.best_score_)

C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of

C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of

{'C': 10.0} 0.9547259729944401


C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [32]:
clf = LinearSVC(C=grid.best_params_['C'], penalty='l1', multi_class='ovr', dual=False)
clf.fit(scX_train, y2_train)
print("train accuaracy: %0.4f" % clf.score(scX_train, y2_train))
print("test accuaracy: %0.4f" % clf.score(scX_test, y2_test))

train accuaracy: 0.9585
test accuaracy: 0.9416


C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [33]:
y2_pred = clf.predict(scX_test)
h2 = hamming_loss(y2_test, y2_pred)
print(h2)

0.058360352014821676


In [34]:
# 3. Species (L1-penalized SVM)
# Determine C by 10 fold CV
grid.fit(scX_train, y3_train)
print(grid.best_params_, grid.best_score_)

C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of

C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of

{'C': 1.0} 0.9618745035742653


C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [35]:
clf = LinearSVC(C=grid.best_params_['C'], penalty='l1', multi_class='ovr', dual=False)
clf.fit(scX_train, y3_train)
print("train accuaracy: %0.4f" % clf.score(scX_train, y3_train))
print("test accuaracy: %0.4f" % clf.score(scX_test, y3_test))

train accuaracy: 0.9658
test accuaracy: 0.9467


C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [36]:
y3_pred = clf.predict(scX_test)
h3 = hamming_loss(y3_test, y3_pred)
print(h3)

0.053265400648448355


In [37]:
# exact match

count = 0
for i in range(len(y1_test)):
    if y1_pred[i]==y1_test.iloc[i] and \
    y2_pred[i]==y2_test.iloc[i] and \
    y3_pred[i]==y3_test.iloc[i]:
        count += 1

e =(count*100)/len(y1_test) # exact math
h=(h1+h2+h3)/3 # hamming loss

print("exact match: %0.3f" % e, "%")        
print("hamming loss: %0.3f" % h)

exact match: 89.764 %
hamming loss: 0.062


1. (b) iv. Repeat 1(b)iii by using SMOTE or any other method you know to remedy class imbalance. Report your conclusions about the classifiers you trained.

In [38]:
# making string classes to numerical classes for smote
from sklearn.preprocessing import LabelEncoder

# converting training label
lb1 = LabelEncoder().fit_transform(y1_train)
lb2 = LabelEncoder().fit_transform(y2_train)
lb3 = LabelEncoder().fit_transform(y3_train)
# converting test label
te1 = LabelEncoder().fit_transform(y1_test)
te2 = LabelEncoder().fit_transform(y2_test)
te3 = LabelEncoder().fit_transform(y3_test)

In [39]:
# smote
from imblearn.over_sampling import SMOTE

X1_smote, y1_smote = SMOTE().fit_resample(scX_train, lb1)
X2_smote, y2_smote = SMOTE().fit_resample(scX_train, lb2)
X3_smote, y3_smote = SMOTE().fit_resample(scX_train, lb3)

In [40]:
# 1. Family (L1-penalized SVM)
# Determine C by 10 fold CV
grid.fit(X1_smote, y1_smote)
print(grid.best_params_, grid.best_score_)

C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of

C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of

{'C': 10000.0} 0.9524699382515437


C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [41]:
clf = LinearSVC(C=grid.best_params_['C'], penalty='l1', multi_class='ovr', dual=False)
clf.fit(X1_smote, y1_smote)
print("train accuaracy: %0.4f" % clf.score(X1_smote, y1_smote))
print("test accuaracy: %0.4f" % clf.score(scX_test, te1))

train accuaracy: 0.9532
test accuaracy: 0.9097


C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [42]:
y1_pred = clf.predict(scX_test)
h1 = hamming_loss(te1, y1_pred)
print(h1)

0.09031959240389069


In [43]:
# 2. Genus (L1-penalized SVM)
# Determine C by 10 fold CV
grid.fit(X2_smote, y2_smote)
print(grid.best_params_, grid.best_score_)

C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of

C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of

{'C': 100.0} 0.9573676928398478


C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [44]:
clf = LinearSVC(C=grid.best_params_['C'], penalty='l1', multi_class='ovr', dual=False)
clf.fit(X2_smote, y2_smote)
print("train accuaracy: %0.4f" % clf.score(X2_smote, y2_smote))
print("test accuaracy: %0.4f" % clf.score(scX_test, te2))

train accuaracy: 0.9584
test accuaracy: 0.9078


C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [45]:
y2_pred = clf.predict(scX_test)
h2 = hamming_loss(te2, y2_pred)
print(h2)

0.0921723019916628


In [46]:
# 3. Species (L1-penalized SVM)
# Determine C by 10 fold CV
grid.fit(X3_smote, y3_smote)
print(grid.best_params_, grid.best_score_)

C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of

C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of

{'C': 10.0} 0.9617502054231717


C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [48]:
clf = LinearSVC(C=grid.best_params_['C'], penalty='l1', multi_class='ovr', dual=False)
clf.fit(X3_smote, y3_smote)
print("train accuaracy: %0.4f" % clf.score(X3_smote, y3_smote))
print("test accuaracy: %0.4f" % clf.score(scX_test, te3))

train accuaracy: 0.9625
test accuaracy: 0.9546


C:\Users\Myunghee\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [49]:
y3_pred = clf.predict(scX_test)
h3 = hamming_loss(te3, y3_pred)
print(h3)

0.04539138490041686


In [50]:
# exact match

count = 0
for i in range(len(y1_test)):
    if y1_pred[i]==te1[i] and \
    y2_pred[i]==te2[i] and \
    y3_pred[i]==te3[i]:
        count += 1

h=(h1+h2+h3)/3
e =(count*100)/len(y1_test)
print("hamming loss: %0.3f" % h)
print("exact match: %0.3f" % e, "%")

hamming loss: 0.076
exact match: 85.410 %


2. K-Means Clustering on a Multi-Class and Multi-Label Data Set Monte-Carlo Simulation: 

Perform the following procedures 50 times, and report the average and standard deviation of the 50 Hamming Distances that you calculate.

(a) Use k-means clustering on the whole Anuran Calls (MFCCs) Data Set (do not split the data into train and test, as we are not performing supervised learning in this exercise). Choose k ∈ {1,2,...,50} automatically based on one of the methods provided in the slides (CH or Gap Statistics or scree plots or Silhouettes) or any other method you know. 

In [51]:
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
# calinski_harabaz_score (CH index)

# selecting K among {1, 2, ..., 50} 
def Kmeans_k (X):
    dic = dict()
    for i in range(49):
        k = i+2
        kmeans = KMeans(n_clusters=k).fit(X)
        labels= kmeans.labels_
        
        s_avg = silhouette_score(X, labels)
        dic[k]=s_avg
    k_select=pd.DataFrame(dic, index=['s_avg']).T
    k_select = k_select.sort_values(by=['s_avg'], ascending = False)
    K = k_select.iloc[0].name
    return K

In [53]:
K = Kmeans_k(X)
print(K)

4


2. (b) In each cluster, determine which family is the majority by reading the true labels. Repeat for genus and species. 

(c) Now for each cluster you have a majority label triplet (family, genus, species). Calculate the average Hamming distance, Hamming score, and Hamming loss5 between the true labels and the labels assigned by clusters.

In [54]:
# making string classes to numerical classes
family = df.loc[:, 'Family']
genus = df.loc[:, 'Genus']
species = df.loc[:, 'Species']
l1=LabelEncoder().fit(family)
l2=LabelEncoder().fit(genus)
l3=LabelEncoder().fit(species)
label1 = l1.transform(family)
label2 = l2.transform(genus)
label3 = l3.transform(species)

In [57]:
tot_table=dict()

# repeating 50 times K-means clustering (K=4)
for i in range(50):
    kmeans = KMeans(n_clusters=K).fit(X)
    labels= kmeans.labels_
    cl_table = pd.DataFrame(labels, columns=['cluster'])
    cl_table.insert(1,"family", label1)
    cl_table.insert(2,"genus", label2)
    cl_table.insert(3,"species", label3)
    cl_table = cl_table.sort_values(by=['cluster'])
    
    clust=cl_table.iloc[:,0]
    c = np.bincount(clust) # counting # of each cluster

    # splitting data according to clusters
    clust0=cl_table.iloc[:c[0],:]  # cluster 0
    clust1=cl_table.iloc[c[0]:c[0]+c[1],:] # cluster 1
    clust2=cl_table.iloc[c[0]+c[1]:c[0]+c[1]+c[2],:] # cluster 2
    clust3=cl_table.iloc[c[0]+c[1]+c[2]:,:] # cluster 3

    dic_class=dict()
    dic_loss=dict()
    
    # each cluster with a majority label triplet(family, genus, species)
    for j in range(3):
        c0=np.bincount(clust0.iloc[:,j+1])
        c1=np.bincount(clust1.iloc[:,j+1])
        c2=np.bincount(clust2.iloc[:,j+1])
        c3=np.bincount(clust3.iloc[:,j+1])
        m0=np.argmax(c0) # majority label of cluster 0
        m1=np.argmax(c1) # majority label of cluster 1
        m2=np.argmax(c2) # majority label of cluster 2
        m3=np.argmax(c3) # majority label of cluster 3
    
        dic_class[j+1]=m0, m1, m2, m3
        # the # of woringly assigned labels 
        dic_loss[j+1]=len(clust)-(c0[m0]+c1[m1]+c2[m2]+c3[m3])
        
    # assigned family labels list for each cluster
    fam=l1.inverse_transform(dic_class[1])
    # assigned genus labels list for each cluster
    gen=l2.inverse_transform(dic_class[2])
    # assigned species labels list for each cluster
    spe=l3.inverse_transform(dic_class[3])
    
    # hamming loss
    HL=(dic_loss[1]+dic_loss[2]+dic_loss[3])/(len(clust)*3)
    
    # each 50 iteration, family, genus, species labels for each cluster
    # and hamming loss
    tot_table[i]=fam, gen, spe, HL

In [64]:
# answers of 2. (b) and (c)
T=pd.DataFrame(tot_table)
print(T)

                                                  0   \
0  [Hylidae, Leptodactylidae, Dendrobatidae, Hyli...   
1        [Hypsiboas, Adenomera, Ameerega, Hypsiboas]   
2  [HypsiboasCinerascens, AdenomeraHylaedactylus,...   
3                                           0.222423   

                                                  1   \
0  [Dendrobatidae, Hylidae, Hylidae, Leptodactyli...   
1        [Ameerega, Hypsiboas, Hypsiboas, Adenomera]   
2  [Ameeregatrivittata, HypsiboasCinerascens, Hyp...   
3                                           0.222423   

                                                  2   \
0  [Hylidae, Leptodactylidae, Dendrobatidae, Hyli...   
1        [Hypsiboas, Adenomera, Ameerega, Hypsiboas]   
2  [HypsiboasCinerascens, AdenomeraHylaedactylus,...   
3                                           0.222423   

                                                  3   \
0  [Leptodactylidae, Dendrobatidae, Hylidae, Hyli...   
1        [Adenomera, Ameerega, Hypsiboas, Hyp

Report the average and standard deviation of the 50 Hamming Distances that you calculate.

(Answer) 
Hamming score is the fraction of correctly classifed labels to the total number of labels. 

Hamming loss is the fraction of wrongly classified labels to the total number of labels. Thus, hamming score is '1-Hamming loss.'

Hamming distance is the fraction of wrongly classified labels to the total number of samples. Thus, hamming distance is "Hamming loss X 'the number of labels(in this case: 3).'

Hamming score and hamming distance can be easily calculated from hamming loss, so I calculated only hamming loss in this HW. 

In [65]:
import statistics

m = statistics.mean(T.iloc[3, :])
s = np.std(T.iloc[3, :])

print("avearge: %0.3f" % m)
print("standard deviation: %0.3f" % s)

avearge: 0.225
standard deviation: 0.007
